openml/datasets/functions.py

```diff
@@ -359,6 +361,57 @@ def get_dataset(dataset_id):
     return dataset
 
 
+def attributes_arff_from_df(df):
+    """Create the attributes as specified by the ARFF format using a dataframe.
+
+    Arguments:
+        df : DataFrame, shape (n_samples, n_features)
+            The dataframe containing the data set.
+
+    Returns:
+        attributes_arff : str
+            The data set attributes as required by the ARFF format.
+    """
+    PD_DTYPES_TO_ARFF_DTYPE = {
+        'integer': 'INTEGER',
+        'floating': 'REAL',
+        'string': 'STRING'
+    }
+    attributes_arff = []
+    for column_name in df:
+        # skipna=True does not infer properly the dtype. The NA values are
+        # dropped before the inference instead.
+        column_dtype = pd.api.types.infer_dtype(df[column_name].dropna())
+
+        if column_dtype == 'categorical':
+            # for categorical feature, arff expects a list string. However, a
+            # categorical column can contain mixed type and we should therefore
+            # raise an error asking to convert all entries to string.
+            categories = df[column_name].cat.categories
+            categories_dtype = pd.api.types.infer_dtype(categories)
+            if categories_dtype not in ('string', 'unicode'):
+                raise ValueError("The column '{}' of the dataframe is of "
+                                 "'category' dtype. Therefore, all values in "
+                                 "this columns should be string. Please "
+                                 "convert the entries which are not string. "
+                                 "Got {} dtype in this column."
+                                 .format(column_name, categories_dtype))
+            attributes_arff.append((column_name, categories.tolist()))
+        elif column_dtype == 'boolean':
```

Comment:
why not use dtype=bool?

Reply:
because we are using the `pd.api.types.infer_dtype` function which return string.